yarn-project/noir-contracts/src/contracts/inclusion_proofs_contract/src/main.nr

```diff
@@ -0,0 +1,230 @@
+            PUBLIC_DATA_TREE_HEIGHT,
+            GENERATOR_INDEX__PUBLIC_LEAF_INDEX,
+        },
+        oracle::{
+            get_block_data::get_block_data,
+            get_membership_witness::{
+                get_membership_witness,
+                MembershipWitness,
+            },
+            get_sibling_path::get_sibling_path,
+            get_low_nullifier_membership_witness::{
+                get_low_nullifier_membership_witness,
+                LowNullifierMembershipWitness,
+            },
+        },
+        hash::pedersen_hash,
+    };
+    use dep::value_note::value_note::{ValueNote, ValueNoteMethods, VALUE_NOTE_LEN};
+
+    use crate::utils::{
+        full_field_less_than,
+        full_field_greater_than,
+    };
+
+    struct Storage {
+        private_values: Map<Set<ValueNote, VALUE_NOTE_LEN>>,
+        public_value: PublicState<Field, 1>,
+    }
+
+    impl Storage {
+        fn init(context: Context) -> Self {
+            Storage {
+                private_values: Map::new(
+                    context,
+                    1, // Storage slot
+                    |context, slot| {
+                        Set::new(context, slot, ValueNoteMethods)
+                    },
+                ),
+                public_value: PublicState::new(
+                    context,
+                    2, // Storage slot
+                    FieldSerializationMethods,
+                ),
+            }
+        }
+    }
+
+    #[aztec(private)]
+    fn constructor(public_value: Field) {
+        let selector = compute_selector("_initialize(Field)");
+        context.call_public_function(context.this_address(), selector, [public_value]);
+    }
+
+    #[aztec(public)]
+    internal fn _initialize(value: Field) {
+        storage.public_value.write(value);
+    }
+
+    // Creates a value note owned by `owner`.
+    #[aztec(private)]
+    fn create_note(
+        owner: AztecAddress,
+        value: Field,
+    ) {
+        let owner_private_values = storage.private_values.at(owner.address);
+        let mut note = ValueNote::new(value, owner.address);
+        owner_private_values.insert(&mut note, true);
+    }
+
+
+    // Proves that the owner owned a ValueNote at block `block_number`.
+    #[aztec(private)]
+    fn proveNoteInclusion(
+        owner: AztecAddress,
+        block_number: Field, // The block at which we'll prove that the note exists
+    ) {
+        // TODO: assert that block number is less than the block number of context.block_data
+        // --> This will either require a new oracle method that returns block_data.global_variables_hash preimage
+        //     or modifying the private context so that we somehow expose it.
+
+        // 1) Get historic block data from oracle and ensure that the block hash is included in the current blocks tree
+        //    root.
+        let block_data = get_block_data(block_number, context.block_data.blocks_tree_root);
+
+        // 2) Get the note from PXE.
```

Comment:
The amount of boiler plate for this is quite wild.

Reply:
I think the only thing that isnt boilerplate is the actual commitment itself, this could probably be moved into a method to be reused? Potentially bolted onto the context

Reply:
I think that manually calling the compute_unique_siloed_note_hash function is not common enough to justify bolting it onto context. But getting a first note from a set and unwrapping it seems very common.

Should we expand `Set` to have get_first function? Then this code:
```
        let private_values = storage.private_values.at(owner.address);
        let options = NoteGetterOptions::new().select(1, owner.address).set_limit(1);
        let notes = private_values.get_notes(options);
        let note = notes[0].unwrap();
```

would look like this:
```
        let note = storage.private_values.at(owner.address).get_first();
```

WDYT?

Reply:
A `Set` is by definition an unordered collection. There's no such thing as "first".

Reply:
@iAmMichaelConnor `get_one()`?

I am not really rooting for doing this now as it's not clear yet if this is a common usecase.